identity/src/rsa.rs

```diff
@@ -46,27 +49,85 @@ impl Keypair {
     /// format (i.e. unencrypted) as defined in [RFC5208].
     ///
     /// [RFC5208]: https://tools.ietf.org/html/rfc5208#section-5
+    #[deprecated(
+        since = "0.2.0",
+        note = "This method name is inaccurate, use `Keypair::try_decode_pkcs8` instead."
+    )]
     pub fn from_pkcs8(der: &mut [u8]) -> Result<Keypair, DecodingError> {
         let kp = RsaKeyPair::from_pkcs8(der)
             .map_err(|e| DecodingError::failed_to_parse("RSA PKCS#8 PrivateKeyInfo", e))?;
+        let kp = Keypair {
+            inner: Arc::new(kp),
+            raw_key: der.to_vec(),
+        };
         der.zeroize();
-        Ok(Keypair(Arc::new(kp)))
+        Ok(kp)
     }
 
     /// Get the public key from the keypair.
     pub fn public(&self) -> PublicKey {
-        PublicKey(self.0.public_key().as_ref().to_vec())
+        PublicKey(self.inner.public_key().as_ref().to_vec())
     }
 
     /// Sign a message with this keypair.
     pub fn sign(&self, data: &[u8]) -> Result<Vec<u8>, SigningError> {
-        let mut signature = vec![0; self.0.public_modulus_len()];
+        let mut signature = vec![0; self.inner.public_modulus_len()];
         let rng = SystemRandom::new();
-        match self.0.sign(&RSA_PKCS1_SHA256, &rng, data, &mut signature) {
+        match self
+            .inner
+            .sign(&RSA_PKCS1_SHA256, &rng, data, &mut signature)
+        {
             Ok(()) => Ok(signature),
-            Err(e) => Err(SigningError::new("RSA").source(e)),
+            Err(e) => Err(SigningError::new("RSA", Some(Box::new(e)))),
+        }
+    }
+
+    /// Try to decode an RSA keypair from a DER-encoded private key.
+    /// Note that a copy of the undecoded byte array will be stored for encoding.
+    pub fn try_decode_der(bytes: &mut [u8]) -> Result<Keypair, DecodingError> {
+        match RsaKeyPair::from_der(bytes) {
+            Ok(kp) => {
+                let kp = Self {
+                    inner: Arc::new(kp),
+                    raw_key: bytes.to_vec(),
+                };
+                bytes.zeroize();
+                Ok(kp)
+            }
+            Err(e) => Err(DecodingError::failed_to_parse("RSA", e)),
+        }
+    }
```

Comment:
This represents a problem. If somebody decodes this keypair from DER and then uses `to_raw_bytes`, we will write spec-incompliant bytes to the protobuf because they won't be PXIX encoded. I think we should not add this function.